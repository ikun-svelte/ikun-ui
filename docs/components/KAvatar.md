---
title: KAvatar
lang: en-US
---

# KAvatar

Avatars can be used to represent people or objects. It supports images, Icons, or characters.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/avatar
```

```bash [yarn]
yarn add @ikun-ui/avatar
```

```bash [npm]
npm install @ikun-ui/avatar
```

:::

## Basic usage

Use `size` and `radius` prop to set avatar's size and radius.

<demo src="avatar/basic.svelte"  github='Avatar'></demo>

## Types

It supports images, Icons, or characters.

<demo src="avatar/types.svelte" github='Avatar'></demo>

## Fallback

fallback when image load error.

<demo src="avatar/fallback.svelte" github='Avatar'></demo>

## Fit Container

Set how the image fit its container for an image avatar, same as [object-fit <span class="i-carbon-link text-12px" />](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit).

<demo src="avatar/fit-container.svelte" github='Avatar'></demo>

## Avatar Props

| Name   | Type                                                 | Default | Description                                                |
| ------ | ---------------------------------------------------- | ------- | ---------------------------------------------------------- |
| icon   | `string`                                             | `''`    | The class name of the icon, following the unocss standard. |
| src    | `string`                                             | `''`    | the source of the image for an image avatar..              |
| srcSet | `string`                                             | `''`    | native attribute `srcset` of image avatar.                 |
| alt    | `string`                                             | `''`    | native attribute `alt` of image avatar.                    |
| fit    | `fill` / `contain` / `cover` / `none` / `scale-down` | `cover` | set how the image fit its container for an image avatar.   |
| size   | `string` / `number`                                  | `50px`  | avatar size, the default unit is `px`.                     |
| radius | `string` / `number`                                  | `4px`   | avatar radius, the default unit is `px`.                   |
| cls    | `string`                                             | `''`    | Additional class for component                             |
| attrs  | `Record<string, string>`                             | `{}`    | Additional attributes                                      |

## Avatar Events

| Name  | Description                    | Type                          |
| ----- | ------------------------------ | ----------------------------- |
| error | trigger when image load error. | `(event: CustomEvent)=> void` |

## Avatar Slots

| Name    | Description               |
| ------- | ------------------------- |
| default | Customize default content |
