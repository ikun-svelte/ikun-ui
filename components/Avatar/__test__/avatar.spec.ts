import { afterEach, expect, test, describe, beforeEach } from 'vitest';
import KAvatar from '../src';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KAvatar', () => {
	test('props: cls', async () => {
		const instance = new KAvatar({
			target: host,
			props: {
				cls: 'k-avatar--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-avatar--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: icon', async () => {
		const instance = new KAvatar({
			target: host,
			props: {
				icon: 'i-carbon-settings'
			}
		});
		expect(instance).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('i-carbon-settings k-icon-transition')
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: src', async () => {
		const src = 'https://ikun-ui.netlify.app/logo.svg';
		const instance = new KAvatar({
			target: host,
			props: {
				src
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('img')?.getAttribute('src')).toBe(src);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: fit', async () => {
		const src = 'https://ikun-ui.netlify.app/logo.svg';
		const fits = ['fill', 'contain', 'cover', 'none', 'scale-down'];
		for (const fit of fits) {
			host.remove();
			initHost();
			const instance = new KAvatar({
				target: host,
				props: {
					src,
					fit
				}
			});
			expect(instance).toBeTruthy();
			expect(host.querySelector('img')?.getAttribute('style')).toContain(`object-fit: ${fit};`);
			expect(host.innerHTML).matchSnapshot();
		}
	});

	test('props: size is number', async () => {
		const size = 30;
		const instance = new KAvatar({
			target: host,
			props: {
				size
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('.k-avatar')?.getAttribute('style')).toContain(
			`width: ${size}px; height: ${size}px`
		);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: size is string', async () => {
		const size = '2rem';
		const instance = new KAvatar({
			target: host,
			props: {
				size
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('.k-avatar')?.getAttribute('style')).toContain(
			`width: ${size}; height: ${size}`
		);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: radius is number', async () => {
		const radius = 10;
		const instance = new KAvatar({
			target: host,
			props: {
				radius
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('.k-avatar')?.getAttribute('style')).toContain(
			`border-radius: ${radius}px;`
		);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: radius is string', async () => {
		const radius = '50%';
		const instance = new KAvatar({
			target: host,
			props: {
				radius
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('.k-avatar')?.getAttribute('style')).toContain(
			`border-radius: ${radius};`
		);
		expect(host.innerHTML).matchSnapshot();
	});
});
