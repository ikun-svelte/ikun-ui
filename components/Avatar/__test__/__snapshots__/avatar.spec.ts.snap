// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KAvatar > props: cls 1`] = `"<div class=\\"k-avatar k-avatar--test\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"></div>"`;

exports[`Test: KAvatar > props: fit 1`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"><img src=\\"https://ikun-ui.netlify.app/logo.svg\\" alt=\\"\\" style=\\"object-fit: fill;\\"></div>"`;

exports[`Test: KAvatar > props: fit 2`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"><img src=\\"https://ikun-ui.netlify.app/logo.svg\\" alt=\\"\\" style=\\"object-fit: contain;\\"></div>"`;

exports[`Test: KAvatar > props: fit 3`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"><img src=\\"https://ikun-ui.netlify.app/logo.svg\\" alt=\\"\\" style=\\"object-fit: cover;\\"></div>"`;

exports[`Test: KAvatar > props: fit 4`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"><img src=\\"https://ikun-ui.netlify.app/logo.svg\\" alt=\\"\\" style=\\"object-fit: none;\\"></div>"`;

exports[`Test: KAvatar > props: fit 5`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"><img src=\\"https://ikun-ui.netlify.app/logo.svg\\" alt=\\"\\" style=\\"object-fit: scale-down;\\"></div>"`;

exports[`Test: KAvatar > props: icon 1`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"><div role=\\"\\" aria-hidden=\\"true\\" class=\\"k-icon--base k-icon--base__dark  \\"><div class=\\"i-carbon-settings k-icon-transition \\" style=\\"width: 24px; height: 24px;\\"></div></div></div>"`;

exports[`Test: KAvatar > props: radius is number 1`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 10px;\\"></div>"`;

exports[`Test: KAvatar > props: radius is string 1`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 50%;\\"></div>"`;

exports[`Test: KAvatar > props: size is number 1`] = `"<div class=\\"k-avatar\\" style=\\"width: 30px; height: 30px; border-radius: 4px;\\"></div>"`;

exports[`Test: KAvatar > props: size is string 1`] = `"<div class=\\"k-avatar\\" style=\\"width: 2rem; height: 2rem; border-radius: 4px;\\"></div>"`;

exports[`Test: KAvatar > props: src 1`] = `"<div class=\\"k-avatar\\" style=\\"width: 50px; height: 50px; border-radius: 4px;\\"><img src=\\"https://ikun-ui.netlify.app/logo.svg\\" alt=\\"\\" style=\\"object-fit: cover;\\"></div>"`;
